diff -Naurp include/avr/sleep.h include/avr/sleep.h
--- include/avr/sleep.h	2013-01-29 13:45:51.000000000 +0530
+++ include/avr/sleep.h	2013-01-29 18:05:21.000000000 +0530
@@ -455,6 +455,8 @@
 || defined(__AVR_ATmega48__) \
 || defined(__AVR_ATmega88__) \
 || defined(__AVR_ATmega168__) \
+|| defined(__AVR_ATmega325P__) \
+|| defined(__AVR_ATmega3250P__) \
 || defined(__AVR_ATmega325__) \
 || defined(__AVR_ATmega3250__) \
 || defined(__AVR_ATmega645__) \
