diff -Naurp include/avr/power.h include/avr/power.h
--- include/avr/power.h	2013-03-15 12:22:13.000000000 +0530
+++ include/avr/power.h	2013-03-15 12:22:26.000000000 +0530
@@ -2171,9 +2171,18 @@ do{ \
 #define power_spi_enable()      (PRR &= (uint8_t)~(1 << PRSPI))
 #define power_spi_disable()     (PRR |= (uint8_t)(1 << PRSPI))
 
+#if defined(__AVR_AT90PWM216__) || defined(__AVR_AT90PWM316__)
+
 #define power_usart_enable()    (PRR &= (uint8_t)~(1 << PRUSART))
 #define power_usart_disable()   (PRR |= (uint8_t)(1 << PRUSART))
 
+#else
+
+#define power_usart_enable()    (PRR &= (uint8_t)~(1 << PRUSART0))
+#define power_usart_disable()   (PRR |= (uint8_t)(1 << PRUSART0))
+
+#endif
+
 #define power_timer0_enable()   (PRR &= (uint8_t)~(1 << PRTIM0))
 #define power_timer0_disable()  (PRR |= (uint8_t)(1 << PRTIM0))
 
@@ -2192,9 +2201,18 @@ do{ \
 #define power_psc2_enable()     (PRR &= (uint8_t)~(1 << PRPSC2))
 #define power_psc2_disable()    (PRR |= (uint8_t)(1 << PRPSC2))
 
+#if defined(__AVR_AT90PWM216__) || defined(__AVR_AT90PWM316__)
+
 #define power_all_enable()      (PRR &= (uint8_t)~((1<<PRADC)|(1<<PRSPI)|(1<<PRUSART)|(1<<PRTIM0)|(1<<PRTIM1)|(1<<PRPSC0)|(1<<PRPSC1)|(1<<PRPSC2)))
 #define power_all_disable()     (PRR |= (uint8_t)((1<<PRADC)|(1<<PRSPI)|(1<<PRUSART)|(1<<PRTIM0)|(1<<PRTIM1)|(1<<PRPSC0)|(1<<PRPSC1)|(1<<PRPSC2)))
 
+#else
+
+#define power_all_enable()      (PRR &= (uint8_t)~((1<<PRADC)|(1<<PRSPI)|(1<<PRUSART0)|(1<<PRTIM0)|(1<<PRTIM1)|(1<<PRPSC0)|(1<<PRPSC1)|(1<<PRPSC2)))
+#define power_all_disable()     (PRR |= (uint8_t)((1<<PRADC)|(1<<PRSPI)|(1<<PRUSART0)|(1<<PRTIM0)|(1<<PRTIM1)|(1<<PRPSC0)|(1<<PRPSC1)|(1<<PRPSC2)))
+
+#endif
+
 
 #elif defined(__AVR_AT90PWM81__) \
 || defined(__AVR_AT90PWM161__)
@@ -2226,9 +2244,11 @@ do{ \
 || defined(__AVR_ATmega165PA__) \
 || defined(__AVR_ATmega325__) \
 || defined(__AVR_ATmega325A__) \
+|| defined(__AVR_ATmega325P__) \
 || defined(__AVR_ATmega325PA__) \
 || defined(__AVR_ATmega3250__) \
 || defined(__AVR_ATmega3250A__) \
+|| defined(__AVR_ATmega3250P__) \
 || defined(__AVR_ATmega3250PA__) \
 || defined(__AVR_ATmega645__) \
 || defined(__AVR_ATmega645A__) \
@@ -2389,8 +2409,6 @@ do{ \
 || defined(__AVR_ATmega168PA__) \
 || defined(__AVR_ATmega328__) \
 || defined(__AVR_ATmega328P__) \
-|| defined(__AVR_ATtiny48__) \
-|| defined(__AVR_ATtiny88__) \
 || defined(__AVR_ATtiny828__)
 
 #define power_adc_enable()      (PRR &= (uint8_t)~(1 << PRADC))
@@ -2430,6 +2448,28 @@ do{ \
 
 #endif
 
+#elif defined(__AVR_ATtiny48__) \
+|| defined(__AVR_ATtiny88__)
+
+#define power_adc_enable()      (PRR &= (uint8_t)~(1 << PRADC))
+#define power_adc_disable()     (PRR |= (uint8_t)(1 << PRADC))
+
+#define power_spi_enable()      (PRR &= (uint8_t)~(1 << PRSPI))
+#define power_spi_disable()     (PRR |= (uint8_t)(1 << PRSPI))
+
+#define power_timer0_enable()   (PRR &= (uint8_t)~(1 << PRTIM0))
+#define power_timer0_disable()  (PRR |= (uint8_t)(1 << PRTIM0))
+
+#define power_timer1_enable()   (PRR &= (uint8_t)~(1 << PRTIM1))
+#define power_timer1_disable()  (PRR |= (uint8_t)(1 << PRTIM1))
+
+#define power_twi_enable()      (PRR &= (uint8_t)~(1 << PRTWI))
+#define power_twi_disable()     (PRR |= (uint8_t)(1 << PRTWI))
+
+#define power_all_enable()      (PRR &= (uint8_t)~((1<<PRADC)|(1<<PRSPI)|(1<<PRTIM0)|(1<<PRTIM1)|(1<<PRTWI)))
+#define power_all_disable()     (PRR |= (uint8_t)((1<<PRADC)|(1<<PRSPI)|(1<<PRTIM0)|(1<<PRTIM1)|(1<<PRTWI)))
+
+  
 #elif defined(__AVR_ATtiny24__) \
 || defined(__AVR_ATtiny24A__) \
 || defined(__AVR_ATtiny44__) \
@@ -2639,7 +2679,8 @@ do{ \
 
 #endif
 
-#elif defined(__AVR_ATmega64HVE2__) 
+#elif defined(__AVR_ATmega64HVE__) \
+|| defined(__AVR_ATmega64HVE2__) 
 
 #define power_lin_enable()      (PRR0 &= (uint8_t)~(1 << PRLIN))
 #define power_lin_disable()     (PRR0 |= (uint8_t)(1 << PRLIN))
@@ -2995,9 +3036,11 @@ do{ \
 || defined(__AVR_ATmega324P__) \
 || defined(__AVR_ATmega325__) \
 || defined(__AVR_ATmega325A__) \
+|| defined(__AVR_ATmega325P__) \
 || defined(__AVR_ATmega325PA__) \
 || defined(__AVR_ATmega3250__) \
 || defined(__AVR_ATmega3250A__) \
+|| defined(__AVR_ATmega3250P__) \
 || defined(__AVR_ATmega3250PA__) \
 || defined(__AVR_ATmega328__) \
 || defined(__AVR_ATmega328P__) \
@@ -3143,6 +3186,7 @@ Gets and returns the clock prescaler reg
 
 #elif defined(__AVR_ATmega16HVB__) \
 || defined(__AVR_ATmega16HVBREVB__) \
+|| defined(__AVR_ATmega64HVE__) \
 || defined(__AVR_ATmega32HVB__) \
 || defined(__AVR_ATmega32HVBREVB__) \
 || defined(__AVR_ATmega64HVE2__)
