diff -Naurp configure.ac configure.ac
--- configure.ac	2013-03-15 12:07:15.000000000 +0530
+++ configure.ac	2013-03-15 12:29:03.000000000 +0530
@@ -1153,9 +1153,9 @@ CHECK_AVR_DEVICE(atxmega128a4u)
 AM_CONDITIONAL(HAS_atxmega128a4u, test "x$HAS_atxmega128a4u" = "xyes")
 
 
-# avrtiny10
-CHECK_AVR_DEVICE(avrtiny10)
-AM_CONDITIONAL(HAS_avrtiny10, test "x$HAS_avrtiny10" = "xyes")
+# avrtiny
+CHECK_AVR_DEVICE(avrtiny)
+AM_CONDITIONAL(HAS_avrtiny, test "x$HAS_avrtiny" = "xyes")
 
 CHECK_AVR_DEVICE(attiny4)
 AM_CONDITIONAL(HAS_attiny4, test "x$HAS_attiny4" = "xyes")
@@ -1541,15 +1541,15 @@ AC_CONFIG_FILES([
 ])
 
 
-# avrtiny10
+# avrtiny
 AC_CONFIG_FILES([
-	avr/lib/avrtiny10/Makefile
-	avr/lib/avrtiny10/attiny4/Makefile
-	avr/lib/avrtiny10/attiny5/Makefile
-	avr/lib/avrtiny10/attiny9/Makefile
-	avr/lib/avrtiny10/attiny10/Makefile
-	avr/lib/avrtiny10/attiny20/Makefile
-	avr/lib/avrtiny10/attiny40/Makefile
+	avr/lib/avrtiny/Makefile
+	avr/lib/avrtiny/attiny4/Makefile
+	avr/lib/avrtiny/attiny5/Makefile
+	avr/lib/avrtiny/attiny9/Makefile
+	avr/lib/avrtiny/attiny10/Makefile
+	avr/lib/avrtiny/attiny20/Makefile
+	avr/lib/avrtiny/attiny40/Makefile
 ])
 
 
diff -Naurp devtools/gen-avr-lib-tree.sh devtools/gen-avr-lib-tree.sh
--- devtools/gen-avr-lib-tree.sh	2013-03-15 12:07:15.000000000 +0530
+++ devtools/gen-avr-lib-tree.sh	2013-03-15 12:29:03.000000000 +0530
@@ -351,7 +351,7 @@ atxmega128a1u:crtx128a1u.o:${DEV_DEFS}:$
 atxmega128a4u:crtx128a4u.o:${DEV_DEFS}:${CFLAGS_BIG_MEMORY}:${DEV_ASFLAGS}\
 "
 
-AVRTINY10_DEV_INFO="\
+AVRTINY_DEV_INFO="\
 attiny4:crttn4.o:${DEV_DEFS}:${CFLAGS_SPACE}:${DEV_ASFLAGS};\
 attiny5:crttn5.o:${DEV_DEFS}:${CFLAGS_SPACE}:${DEV_ASFLAGS};\
 attiny9:crttn9.o:${DEV_DEFS}:${CFLAGS_SPACE}:${DEV_ASFLAGS};\
@@ -379,7 +379,7 @@ avrxmega4:AVRXMEGA4_DEV_INFO:${LIB_DEFS}
 avrxmega5:AVRXMEGA5_DEV_INFO:${LIB_DEFS}:${CFLAGS_BIG_MEMORY}:${DEV_ASFLAGS};\
 avrxmega6:AVRXMEGA6_DEV_INFO:${LIB_DEFS}:${CFLAGS_BIG_MEMORY}:${DEV_ASFLAGS};\
 avrxmega7:AVRXMEGA7_DEV_INFO:${LIB_DEFS}:${CFLAGS_BIG_MEMORY}:${DEV_ASFLAGS};\
-avrtiny10:AVRTINY10_DEV_INFO:${LIB_DEFS}:${CFLAGS_SPACE}:${DEV_ASFLAGS}\
+avrtiny:AVRTINY_DEV_INFO:${LIB_DEFS}:${CFLAGS_SPACE}:${DEV_ASFLAGS}\
 "
 
 echo "Generating source directories:"
